### PR TITLE
New fxa client [firefox-android: bendk/new-fxa-client]

### PR DESCRIPTION
**Pull request by bendk**
### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+     *
+     * This is the coroutineContext passed in to the constructor plus a SupervisorJob.  The
+     * SupervisorJob ensures that if one task fails the other still run.
+     *
+     * Because we use a single CoroutineContext, [close] can cancel all active jobs.
+     *
+     * This is public so applications can use the same context to run their jobs in.  For example
+     * the android-components FxaManager uses it for its startup tasks.
+     */
+    public val coroutineContext = coroutineContext + SupervisorJob()
```
Comment (bendk) [inline components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt:38]: Someone with Kotlin experience please double check that I'm doing the correct thing here.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+    fun shouldRecheckAuthStatus(): Boolean {
+        val elasped = (System.currentTimeMillis() - lastAuthCheck).milliseconds
+        if (elasped > 60.seconds) {
+            lastAuthCheck = System.currentTimeMillis()
+            return true
+        } else {
+            return false
+        }
+    }
```
Comment (bendk) [inline components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt:314]: I like this class because it isolates the retry logic, but I don't really know what the correct logic is.

The old android-components code performed 3 network retries for every operation, which seemed quite a lot to me.

Auth checking had a complicated system that boiled down to don't do it too much and definitely don't get stuck in an infinite loop.  Infinite loops shouldn't be an issue anymore since we don't queue a new action to handle the auth check.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+// This allows `adb logcat | grep fxa_client` to pick up all log items from this component
+internal const val LOG_TAG = "fxa_client"
+
+/**
+ * Firefox account client
+ */
+class FxaClient private constructor(
+    private val inner: FirefoxAccount,
+    private val persistCallback: FxaPersistCallback,
+    private val eventHandler: FxaEventHandler,
```
Comment (bendk) [inline components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt:24]: Note that this creates a reference cycle between `FxaAccountManager`, `FxaClient`, and `FxaHandler` that goes through the Rust code and therefore will never be broken.  I think this is okay for now, since all of those objects are intended to live forever, but hopefully we can add some sort of weak reference support to side-step this.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+        inner.simulatePermanentAuthTokenIssue()
+    }
+}
+
+/**
+ * Implemented by applications to save the Fxa state
+ */
+interface FxaPersistCallback {
+    fun persist(data: String)
+}
```
Comment (bendk) [inline components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt:416]: Do we have a naming system for our components?  I tended to prefix everything with `Fxa`, since you don't ever import modules in Kotlin.  I'm open to any suggestions though.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+                } else {
+                    throw e
+                }
+            }
+        }
+    }
+
+    private suspend fun handleBeginOAuthFlow(currentState: FxaAuthState.Disconnected, action: FxaAction.BeginOAuthFlow) {
+        handleBeginEitherOAuthFlow(currentState, action.result) {
+            inner.beginOauthFlow(action.scopes.toList(), action.entrypoint, MetricsParams(mapOf()))
```
Comment (bendk) [inline components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt:159]: I'm I handling `MetricsParams` correctly?  I don't really understand how it's supposed to work.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+ *
+ * [FxaActionProcessor] is "inert" on construction and won't process any events.  Call
+ * [runActionProcessorManager] start event processing.  This creates a manager task that monitors
+ * the action handling loop and restarts it if it throws for any reason.
+ *
+ */
+internal class FxaActionProcessor(
+    private val inner: FirefoxAccount,
+    private val eventHandler: FxaEventHandler,
+    private val tryPersistState: () -> Unit,
```
Comment (mhammond) [inline components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt:30]: This name surprises me - from the implementation the intent is that it never throws an exception, which is the opposite of what I'd expect of a function name starting with `try`.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+    /**
+     * Fetches the profile object for the current client either from the existing cached account,
+     * or from the server (requires the client to have access to the profile scope).
+     *
+     * @param ignoreCache Fetch the profile information directly from the server
+     * @return [Profile] representing the user's basic profile info
+     * @throws FxaException.Unauthorized We couldn't find any suitable access token to make that call.
+     * The caller should then start the OAuth Flow again with the "profile" scope.
+     */
+    suspend fun getProfile(ignoreCache: Boolean): Profile = wrapMethodCallAndPersist("getProfile") {
```
Comment (mhammond) [inline components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt:150]: If `ignoreCache` is false and we end up returning a cached value, we probably don't need to persist anything?

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+     * @return String containing the authentication details in JSON format
+     */
+    fun toJSONString(): String {
+        return inner.toJson()
+    }
+
+    /**
+     * Retrieves the list of the connected devices in the current account, including the current one.
+     */
+    suspend fun getDevices(ignoreCache: Boolean = false): Array<Device> = wrapMethodCall("getDevices") {
```
Comment (mhammond) [inline components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt:290]: mildly surprised we don't want to persist the result in a cache? IOW, I'm surprised this is different to `getProfile()`

**Review by mhammond — COMMENTED** [review]
Sorry, but I didn't get to finish this - it's big :) In general it looks good and some of my comments might be missing the point entirely. I'll get back to it tomorrow.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+    initialState: FxaAuthState = FxaAuthState.fromRust(inner.getAuthState()),
+) {
+    private val channel = Channel<FxaAction>(Channel.UNLIMITED)
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt:33]: I think a relatively high bound could be good here and would help diagnose the processor stalling.  I don't think that should happen, but if it did I would expect that the actions would start piling up and even a bound of 100 or so would result in crash reports.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+    fun close() {
+        channel.close()
+    }
+
+    internal suspend fun processChannel() {
+        Log.d(LOG_TAG, "processChannel started: $currentState")
+        for (action in channel) {
+            Log.d(LOG_TAG, "Processing action: $action")
+            processAction(action)
+            Log.d(LOG_TAG, "Action processed: $action")
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt:57]: There's a generic handler one level up in `runActionProcessorManager` that restarts the entire loop.  I figured that maybe could catch some extra exceptions that a handler here wouldn't.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (mhammond) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt:33]: I initially thought that too, but when writing the comment I also thought that we'd probably *never* expect to end up with 100 "actions" in total for a single session, right? ie, would a bound of 100 actually catch this stalling in practice?

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (mhammond) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt:57]: fair enough - maybe a very short comment to that end?

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+
+    // Does what wrapMethodCall does and also ensures that the state is perisisted at the end of the
+    // operation.
+    private suspend fun<T> wrapMethodCallAndPersist(name: String, methodCall: () -> T): T {
+        return try {
+            wrapMethodCall(name, methodCall)
+        } finally {
+            // Use a finally block since we want to perisist the state regardless of if the method
+            // succeeded or not.
+            tryPersistState()
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt:137]: I wrote it up this way because that's how the existing code worked ([for example, getProfile()](https://github.com/mozilla/application-services/blob/main/components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/PersistedFirefoxAccount.kt#L152-L158)).  I think that's intended to address the case where you the client tries to fetch the profile, realizes its auth token is bad so it deletes it, and returns an auth error.  In that case you still want to persist the state even though the action failed.

So I think this should be okay, since that's how the current API works and I don't think we've ever wanted to break it up into smaller chunks.

However, your idea of a transactional-style API is really interesting.  In addition to allowing us to expose smaller operations, it could also handle the serialization issue.  The whole idea of the action / event system is to ensure that we process in the correct order and also that the consumer responds to the results in the correct order.   Another approach would be to have the consumer start a transaction, perform some operations, handle the results, then drop the transaction.  If we enforced only 1 transaction running at a time, then I think this achieves the same result and maybe would be a nicer API to work with.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt (hunk 8 above)
Comment (bendk) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt:290]: Good point.  I just translated what the old code did, but I think it should have been persisting the state.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/StateTypes.kt [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    /**
+     * Client has disconnected
+     *
+     * @property fromAuthIssues client was disconnected because of invalid auth tokens, for
+     *   example because of a password reset on another device
+     * @property connecting is there an OAuth flow in progress?
+     */
+    data class Disconnected(
+        val fromAuthIssues: Boolean = false,
+        val connecting: Boolean = false,
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/StateTypes.kt:211]: I was attracted to the idea that the FxaAuthState subclass defined which actions were possible and the fields were details about the state.  So if your state is `Disconnected` then you can begin an oauth flow, if it's `Connected` you can't.

Maybe it's better to use a flat enum though.

[the remark this answers is not in the document]
Comment (mhammond) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/StateTypes.kt:211]: OTOH, if you want to keep the current mechanism, maybe "not connected" would be a better name - it's a subtle difference, but it avoids the baggage the term "disconnect" carries. OTOH, maybe that's just my baggage :)(although I think it comes from desktop really)

(I'm still not sure though - needing to examine 2 bools in one variant to determine the difference between "user has never even *tried* to connect an account" and "user needs to reenter their password" doesn't really seem ideal.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+ */
+internal class FxaActionProcessor(
+    private val inner: FirefoxAccount,
+    private val eventHandler: FxaEventHandler,
+    private val persistState: () -> Unit,
+    initialState: FxaAuthState = FxaAuthState.fromRust(inner.getAuthState()),
+) {
+    // Set a high bound on the channel so that if our the processChannel() task dies, or is never
+    // started, we should eventually see error reports.
+    private val channel = Channel<FxaAction>(100)
```
Comment (tarikeshaq) [inline components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt:35]: What do you think about defining 100 as a constant?

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt (hunk 5 above)
Comment (tarikeshaq) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt:164]: It gets passed in from the caller - it includes query parameters that the client wants to append to the URL, eg: all the `utm_*` params in https://mozilla.github.io/ecosystem-platform/relying-parties/reference/query-parameters#generic-parameters

**Review by tarikeshaq — CHANGES_REQUESTED** [review]
I had a first pass but I will need another - thanks a lot Ben!

Like @mhammond said this is fairly large and introduces quite a few changes more than just moving A-C into here (may not be a bad thing, just means it's harder to review properly)

I had a few thoughts that I wrote down and my high-level thought is that I'm wondering if we can take this opportunity to write out a short artifact (a doc? diagram?) that's accessible to read about all the state transitions this PR moves over from A-C and introduces here. That will help me reason about it easier than hopping around Kotlin files (and I'm not too familiar with Kotlin so it introduces a possibility for us to miss things)

**Comment by bendk** [comment]
@mhammond Thanks for the review, I rebased in changes based on most of your suggestions.  The changes to `FxaAuthState` were fairly big, so I made a new commit for that.  I'm definitely liking the new (or maybe old) system, if for no other reason that it forced me to test all states in the unit tests and that uncovered some issues in the corner cases.

File: components/fxa-client/src/fxa_client.udl [hunk 14]
```diff
@@ -891,6 +901,12 @@ dictionary Profile {
   boolean is_default_avatar;
 };
 
+[Enum]
+interface FxaRustAuthState {
+  Disconnected(boolean from_auth_issues);
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/src/fxa_client.udl]: Yes it should :)  I saw this comment right after pushing the commit to change things.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+    internal val retryLogic = RetryLogic()
+    internal var currentState = initialState
+
+    @Volatile
+    internal var simulateNetworkErrorFlag: Boolean = false
+
+    // Queue a new action for processing
+    fun queue(action: FxaAction) {
+        // trySend allows this function to be non-suspend and will never fail since the channel size
+        // is UNLIMITED
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt:45]: I'm very confident that it should be.  Mark raised the opposite concern: even if the queue processor dies, are we sure that we will queue up 100 actions and see an exception?  I'm not so sure on that one.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+            inner.beginPairingFlow(action.pairingUrl, action.scopes.toList(), action.entrypoint, MetricsParams(mapOf()))
+        }
+    }
+
+    private suspend fun handleBeginEitherOAuthFlow(currentState: FxaAuthState.Disconnected, result: CompletableDeferred<String?>?, operation: () -> String) {
+        try {
+            val url = withRetry { operation() }
+            persistState()
+            changeState(currentState.copy(connecting = true), FxaAuthStateTransition.OAUTH_STARTED)
+            sendEvent(FxaEvent.BeginOAuthFlow(url))
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt:198]: Good point, I'll add an extra param here.  If a consumer doesn't care than they can just ignore it.

[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt:198]: Actually, I'm not so sure yet.  I still think this is a good point, but I think it applies more to the `FxaEvent.AuthEvent` (formally `StateChanged`).  I was considering adding field for the OAuth flow kind there, but it's not so easy since there can be multiple flows.

Firefox android doesn't need this yet, so I think we should wait on implementing it, maybe alongside a requirement that only one OAuth flow is in progress at any time.

File: components/fxa-client/src/error.rs [hunk 17]
```diff
@@ -103,6 +106,9 @@ pub enum Error {
     #[error("Remote key and local key mismatch")]
     MismatchedKeys,
 
+    #[error("A required scoped key was missing")]
+    ScopedKeyMissing,
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/src/error.rs:110]: Android-components handles each of these separately.  For `Authentication`, it calls `queueAction(CheckAuthorization)`.  For `ScopedKeyMissing` it goes straight to `logoutFromAuthIssues`.  I think that eventually the clients shouldn't need to deal with this distinction and we should be doing all the auth-checking for them, but I'd rather handle that as a follow up.

I'll fix the name and I also realized that I was making that same mistake in the code.  If we have a cached auth info with the key missing, I think we should retry fetching it before giving up and returning the error.

File: components/fxa-client/src/fxa_client.udl [hunk 18]
```diff
@@ -579,6 +582,8 @@ interface FirefoxAccount {
   //        - This must be one of the scopes requested during the signin flow.
   //        - Only a single scope is supported; for multiple scopes request multiple tokens.
   //    - `ttl` - optionally, the time for which the token should be valid, in seconds.
+  //    - `require_scoped_key` - optionally, throw [FxaError.ScopedKeyMissing] if the scoped_key
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/src/fxa_client.udl:586]: Very good point, I'll make this change.

File: components/fxa-client/src/internal/device.rs [hunk 19]
```diff
@@ -329,6 +326,40 @@ impl FirefoxAccount {
     }
 }
 
+impl TryFrom<String> for DeviceCapability {
+    type Error = Error;
+
+    fn try_from(command: String) -> Result<Self> {
+        match command.as_str() {
+            commands::send_tab::COMMAND_NAME => Ok(DeviceCapability::SendTab),
+            _ => Err(Error::UnknownCommand(command)),
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/src/internal/device.rs:335]: The issue is we need to return something for that case.  I don't think Rust will let me  implement `From<String> for Option<DeviceCapability>`

In practice this shouldn't result in an error.  The `From<UpdateDeviceResponse> for LocalDevice` impl just logs a warning and continues.

File: components/fxa-client/src/internal/state_persistence.rs [hunk 20]
```diff
@@ -99,11 +99,12 @@ pub(crate) struct StateV2 {
     #[serde(default)]
     pub(crate) commands_data: HashMap<String, String>,
     #[serde(default)]
-    pub(crate) device_capabilities: HashSet<DeviceCapability>,
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/src/internal/state_persistence.rs:102]: I thought it would be okay since they're mostly just used as a cache for `ensure_capabilities`.  If we end up making an extra device update requests that seems okay.

However, the other use is in `reregister_current_capabilities()` and I realize that we really should not be using `server_local_device_info` at all for that.  In that scenario, we don't care what the server thinks our capabilities are, we want to re-send what we know our capabilities are.

I'm going to add back `device_capabilities` and have that track whatever was last sent to `ensure_capabilities` or `initialize_device`.  Eventually, I think we should track 2 things: the correct `LocalDevice` info that we know we have, and the `LocalDevice` info that the server thinks we have.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt (hunk 5 above)
Comment (bendk) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt]: Ahh okay.  I don't think Android uses this, but I added this as an optional parameter for the begin oauth actions.

**Comment by bendk** [comment]
Pushed another commit to address @tarikeshaq's comments.  I think we're passed the point of rebasing nicely into the individual commits, but I'll squash things up at the end of the review.

I've been fairly aggressively marking comments resolved, feel fere to open any back up that I didn't address properl.

**Review by mavduevskiy — COMMENTED** [review]
I definitely welcome and appreciate the new API for interracting with `FirefoxAccount` :) But the PR is _a bit on the bigger side_. 

- Changes affect both inner logic of Rust components and kotlin wrappers.
- The kotlin change aims to move some logic from A-C while simultaneously refactoring it (portion of the logic from `FxaAccountManager` is being moved into `FxaClient`).
- There are improvements to address older [problems](https://github.com/mozilla-mobile/android-components/issues/8527).
- Some commit comments are more detailed than a bigger than average PR we have on fenix. 

It looks like just a bit too much for a single thing to land – or even review thoroughly. Given there are 2600 lines of code added, my guess would be we might reach close to 500 hundred comments before landing it. Would you consider landing it in bits @bendk? It would definitely give reviewers more confidence. 

Regarding the kotlin side change, I would agree with @tarikeshaq that some form of document (an RFC) describing the approach and the change would be nice. It's easier to grasp the intention behind the change rather than trying to make sense of all the changes. But overall, it does look good.

Also, the change to kotlin wrapper aims to move logic from A-C while simultaneously refactoring it. It's harder to review and to land – it have to land the A-S change first, and then start using it. Did you consider introducing all new wrappers in A-C layer, then moving them into A-S? It could make testing and reviewing it easier.

Just wanted to praise the work you have done. The scope is huge, and the change is very very much welcomed. Just trying to find a good way to land it. Given it's related to account domain, extra cation is welcomed.

File: build.gradle [hunk 21]
```diff
@@ -22,6 +22,7 @@ buildscript {
         ktlint_version = '0.50.0'
         gradle_download_task_version = '5.2.1'
         junit_version = '4.13.2'
+        mockk_version = '1.13.8'
```
[the remark this answers is not in the document]
Comment (bendk) [reply build.gradle:25]: Is `mockito` recommended over `mockk` or should I be avoiding mocking libraries altogether?

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt [hunk 22]
```diff
@@ -0,0 +1,426 @@
...
+import kotlinx.coroutines.withContext
+import kotlin.coroutines.CoroutineContext
+
+// This allows `adb logcat | grep fxa_client` to pick up all log items from this component
+internal const val LOG_TAG = "fxa_client"
+
+/**
+ * Firefox account client
+ */
+class FxaClient private constructor(
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt:20]: There's 2 public constructors: one inputs the `Config` and the other inputs a persisted JSON string instead.  You can't define one in terms of the other.  It's a bit weird, but not something that I thought needed fixing.

**Comment by bendk** [comment]
@mavduevskiy thanks for the review!  Short-term, I'm going to try to update this PR to incorporate your comments.

Long-term, I like the idea of breaking things up more and definitely don't like doing this all in one pass.  Making the changes in A-C first, then moving everything over to A-S sounds great to me.

I think I can break up the changes to the state machine into smaller commits.  However, before I do that, I'd like to get agreement that something like this code would be a good end-state.  My guess is that those small refactor commits won't make sense without an end-goal in mind.  Maybe this is where your suggestion of an RFC would come into play.  I could put together a high-level description of the changes and we could discuss there.

One of the reasons I bundled everything together is to reduce the number of reviews.  But I can see why reviewing a number of smaller commits would actually be easier.

File: build.gradle (hunk 21 above)
[the remark this answers is not in the document]
Comment (jonalmeida) [reply build.gradle:25]: `mockito` is preferable and we've used it in AC without issues for many years now. There were useful methods in the `mockito-kotlin` library that were helpful as well but to avoid adding more dependencies, we [imported only those specific convenience functions](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/support/test/src/main/java/mozilla/components/support/test/Mock.kt#1). 

If you can avoid a mocking library that would be ideal, but that is more of a philosophy than a good reason. 🙂

**Comment by bendk** [comment]
I think everyone agrees with the slower, step-by-step approach.  Let's close this one.  I think the first step is going to be merging these PRs:
* https://github.com/mozilla/application-services/pull/5887
* https://github.com/mozilla/application-services/pull/5887
